src/lib/OpenEXRCore/openexr_decompress.c

```diff
@@ -0,0 +1,288 @@
+/*
+** SPDX-License-Identifier: BSD-3-Clause
+** Copyright Contributors to the OpenEXR Project.
+*/
+
+#include "openexr.h"
+
+#include "openexr_priv_structs.h"
+#include "openexr_priv_xdr.h"
+#include "openexr_priv_memory.h"
+
+#include <limits.h>
+#include <stdlib.h>
+#include <string.h>
+#include <zlib.h>
+#if defined __SSE2__ || (_MSC_VER >= 1300 && !_M_CEE_PURE)
+# define IMF_HAVE_SSE2 1
+# include <emmintrin.h>
+# include <mmintrin.h>
+#endif
+#if defined __SSE4_1__
+# define IMF_HAVE_SSE4_1 1
+# include <smmintrin.h>
+#endif
+
+/**************************************/
+
+#ifdef IMF_HAVE_SSE4_1
+static void
+reconstruct(uint8_t *buf, size_t outSize)
+{
+    static const size_t bytesPerChunk = sizeof(__m128i);
+    const size_t vOutSize = outSize / bytesPerChunk;
+    const __m128i c = _mm_set1_epi8(-128);
+    const __m128i shuffleMask = _mm_set1_epi8(15);
+    __m128i *vBuf, vPrev;
+    uint8_t prev;
+
+    /*
+     * The first element doesn't have its high bit flipped during compression,
+     * so it must not be flipped here.  To make the SIMD loop nice and
+     * uniform, we pre-flip the bit so that the loop will unflip it again.
+     */
+    buf[0] += -128;
+    vBuf = (__m128i *)buf;
+    vPrev = _mm_setzero_si128();
+
+    for ( size_t i = 0; i < vOutSize; ++i )
+    {
+        __m128i d = _mm_add_epi8(_mm_loadu_si128(vBuf), c);
+
+        /* Compute the prefix sum of elements. */
+        d = _mm_add_epi8(d, _mm_slli_si128(d, 1));
+        d = _mm_add_epi8(d, _mm_slli_si128(d, 2));
+        d = _mm_add_epi8(d, _mm_slli_si128(d, 4));
+        d = _mm_add_epi8(d, _mm_slli_si128(d, 8));
+        d = _mm_add_epi8(d, vPrev);
```

Comment:
Maybe this isn't the time and place to bring this up, but perhaps it's about time to kick these direct intrinsics to the curb and instead use wrappers with understandable names and that, inside them, have implementations for SSE, ARM, generic C fallback, etc. As an example, see [OIIO's simd.h](https://github.com/OpenImageIO/oiio/blob/master/src/include/OpenImageIO/simd.h#L6882). 

Yes, I know that my C++ classes are not a good solution for this C code, but the point is that we could make C wrappers these 4-wide ops, with friendly and self-documenting names and a central place to have alternate implementations -- such as we're going to want to get this running fast on the new Macs with ARM processors.

Reply:
good point, I'll have a look